### PR TITLE
mobile app proxy - request auth token from cloud auth api

### PR DESCRIPTION
**Pull request by joeyorlando**
# What this PR does

Related to https://github.com/grafana/oncall-private/issues/2071

See [this conversation](https://raintank-corp.slack.com/archives/C064R17Q1A8/p1706125615995019) for all the context

## Checklist

- [x] Unit, integration, and e2e (if applicable) tests updated
- [x] Documentation added (or `pr:no public docs` PR label added if not required)
- [x] `CHANGELOG.md` updated (or `pr:no changelog` PR label added if not required)

### CODE REVIEW

File: engine/requirements.txt [hunk 1]
```diff
@@ -56,4 +56,3 @@ babel==2.12.1
 drf-spectacular==0.26.5
 grpcio==1.57.0
 markdown2==2.4.10
-PyJWT==2.8.0
```
Comment (joeyorlando) [inline engine/requirements.txt:59]: was only used by the initial mobile-app proxy gateway POC

File: engine/common/oncall_gateway/client.py [hunk 2]
```diff
@@ -44,7 +44,7 @@ def __init__(self, status, url, msg="", method="GET"):
         self.msg = msg
 
     def __str__(self):
-        return f"LabelsRepoAPIException: status={self.status} url={self.url} method={self.method} error={self.msg}"
+        return f"ChatopsProxyAPIException: status={self.status} url={self.url} method={self.method} error={self.msg}"
```
Comment (joeyorlando) [inline engine/common/oncall_gateway/client.py:47]: fixing a random typo I came across

File: engine/common/cloud_auth_api/client.py
Comment (joeyorlando) [inline engine/common/cloud_auth_api/client.py:1]: main focus

File: engine/apps/mobile_app/views.py [hunk 3]
```diff
@@ -167,19 +167,28 @@ def _construct_jwt_payload(cls, user: "User") -> typing.Dict[str, typing.Any]:
...
+        token_claims = cls._construct_token_claims(user)
+        token_scopes = {
+            cls.SupportedDownstreamBackends.INCIDENT: [CloudAuthApiClient.Scopes.INCIDENT_WRITE],
+        }[downstream_backend]
+
+        org = user.organization
+        stack_id = org.stack_id
+        token = CloudAuthApiClient().request_signed_token(org.org_id, stack_id, token_scopes, token_claims)
+
+        return f"{stack_id}:{token}"
```
Comment (joeyorlando) [inline engine/apps/mobile_app/views.py:184]: main focus

File: engine/apps/mobile_app/views.py [hunk 4]
```diff
@@ -217,13 +228,14 @@ def _proxy_request(self, request: Request, *args, **kwargs) -> Response:
...
                 params=request.query_params.dict(),
-                headers=self._get_downstream_headers(request, user),
+                headers=self._get_downstream_headers(request, downstream_backend, user),
             )
 
             return Response(status=downstream_response.status_code, data=downstream_response.json())
         except (
             requests.exceptions.RequestException,
             requests.exceptions.JSONDecodeError,
+            CloudAuthApiException,
```
Comment (joeyorlando) [inline engine/apps/mobile_app/views.py:238]: if the cloud auth API's `POST /v1/sign` endpoint returns anything that is non-200, we return HTTP 502 to the client (aka the mobile app)

File: engine/apps/mobile_app/views.py [hunk 5]
```diff
@@ -135,12 +136,10 @@ class MobileAppGatewayView(APIView):
...
     permission_classes = (IsAuthenticated,)
 
-    class SupportedDownstreamBackends:
+    class SupportedDownstreamBackends(enum.StrEnum):
         INCIDENT = "incident"
 
-    ALL_SUPPORTED_DOWNSTREAM_BACKENDS = [
-        SupportedDownstreamBackends.INCIDENT,
-    ]
+    ALL_SUPPORTED_DOWNSTREAM_BACKENDS = list(SupportedDownstreamBackends)
```
Comment (joeyorlando) [inline engine/apps/mobile_app/views.py:142]: [TIL](https://docs.python.org/3/library/enum.html#enum.EnumType.__contains__)

**Review by mderynck — APPROVED** [review]

**Pull request merged**